src/irr/asset/CGLSLVirtualTexturingBuiltinIncludeLoader.h

```diff
@@ -0,0 +1,368 @@
+	return vec3(vec2(pageXY),float(pageID>>addr_layer_shift[formatID]));
+}
+uvec2 irr_glsl_unpackWrapModes(in uvec2 texData)
+{
+    return (texData>>uvec2(28u,30u)) & uvec2(0x03u);
+}
+uint irr_glsl_unpackMaxMipInVT(in uvec2 texData)
+{
+    return (texData.y>>24)&0x0fu;
+}
+vec3 irr_glsl_unpackVirtualUV(in uvec2 texData)
+{
+	// assert that PAGE_SZ_LOG2<8 , or change the line to uvec3(texData.yy<<uvec2(PAGE_SZ_LOG2,PAGE_SZ_LOG2-8u),texData.y>>16u)
+    uvec3 unnormCoords = uvec3(texData.y<<PAGE_SZ_LOG2,texData.yy>>uvec2(8u-PAGE_SZ_LOG2,16u))&uvec3(uvec2(0xffu)<<PAGE_SZ_LOG2,0xffu);
+    return vec3(unnormCoords);
+}
+vec2 irr_glsl_unpackSize(in uvec2 texData)
+{
+	return vec2(texData.x&0xffffu,texData.x>>16u);
+}
+)";
+		s += //those are supposed to be used by VT fucntions only (hence extra VT_ pefix)
+			"\n#define irr_glsl_VT_getPgTabSzLog2 " + args[ix_get_pgtab_sz_log2_name] +
+			"\n#define irr_glsl_VT_getPhysPgTexSzRcp " + args[ix_get_phys_pg_tex_sz_rcp_name] +
+			"\n#define irr_glsl_VT_getVTexSzRcp " + args[ix_get_vtex_sz_rcp_name] +
+			"\n#define irr_glsl_VT_layer2pid " + args[ix_get_layer2pid_name]
+			;
+        s += R"(
+vec3 irr_glsl_vTextureGrad_helper(in uint formatID, in vec3 virtualUV, in int clippedLoD, in int levelInTail)
+{
+    uvec2 pageID = textureLod(pageTable,virtualUV,clippedLoD).xy;
+
+	const uint pageTableSizeLog2 = irr_glsl_VT_getPgTabSzLog2(formatID);
+    const float phys_pg_tex_sz_rcp = irr_glsl_VT_getPhysPgTexSzRcp(formatID);
+	// assert that pageTableSizeLog2<23
+
+	// this will work because pageTables are always square and PoT and IEEE754
+	uint thisLevelTableSize = (pageTableSizeLog2-uint(clippedLoD))<<23;
+
+	vec2 tileCoordinate = uintBitsToFloat(floatBitsToUint(virtualUV.xy)+thisLevelTableSize);
+	tileCoordinate = fract(tileCoordinate); // optimize this fract at some point
+	tileCoordinate = uintBitsToFloat(floatBitsToUint(tileCoordinate)+uint((PAGE_SZ_LOG2-levelInTail)<<23));
+    tileCoordinate += packingOffsets[levelInTail];
+	tileCoordinate += vec2(TILE_PADDING,TILE_PADDING);
+	tileCoordinate *= phys_pg_tex_sz_rcp;
+
+	vec3 physicalUV = irr_glsl_unpackPageID(levelInTail!=0 ? pageID.y:pageID.x, formatID);
+	physicalUV.xy *= vec2(PAGE_SZ+2*TILE_PADDING)*phys_pg_tex_sz_rcp;
+
+	// add the in-tile coordinate
+	physicalUV.xy += tileCoordinate;
+    return physicalUV;
+}
+
+float irr_glsl_lengthManhattan(vec2 v)
+{
+	v = abs(v);
+    return v.x+v.y;
+}
+float irr_glsl_lengthSq(in vec2 v)
+{
+  return dot(v,v);
+}
+// textureGrad emulation
+vec4 irr_glsl_vTextureGrad(in uint formatID, in vec3 virtualUV, in mat2 dOriginalScaledUV, in int originalMaxFullMip)
+{
+	// returns what would have been `textureGrad(originalTexture,gOriginalUV[0],gOriginalUV[1])
+	// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/chap15.html#textures-normalized-operations
+	const float kMaxAnisotropy = float(2u*TILE_PADDING);
+	// you can use an approx `log2` if you know one
+#if APPROXIMATE_FOOTPRINT_CALC
```

Comment:
change the name of this define into something prefixed and indicating semantics clearly